cdk/PgBouncer.ts

```diff
@@ -0,0 +1,179 @@
+import {
+  aws_ec2 as ec2,
+  aws_iam as iam,
+  aws_secretsmanager as secretsmanager,
+  Stack,
+} from "aws-cdk-lib";
+import { Construct } from "constructs";
+
+import * as fs from "fs";
+import * as path from "path";
+
+export interface PgBouncerProps {
+  /**
+   * Name for the pgbouncer instance
+   */
+  instanceName: string;
+
+  /**
+   * VPC to deploy PgBouncer into
+   */
+  vpc: ec2.IVpc;
+
+  /**
+   * The RDS instance to connect to
+   */
+  database: {
+    connections: ec2.Connections;
+    secret: secretsmanager.ISecret;
+  };
+
+  /**
+   * Whether to deploy in public subnet
+   * @default false
+   */
+  usePublicSubnet?: boolean;
+
+  /**
+   * Instance type for PgBouncer
+   * @default t3.small
+   */
+  instanceType?: ec2.InstanceType;
+
+  /**
+   * PgBouncer configuration options
+   */
+  pgBouncerConfig: {
+    poolMode: "transaction" | "session" | "statement";
+    maxClientConn: number;
+    defaultPoolSize: number;
+    minPoolSize: number;
+    reservePoolSize: number;
+    reservePoolTimeout: number;
+    maxDbConnections: number;
+    maxUserConnections: number;
+  };
+}
+
+export class PgBouncer extends Construct {
+  public readonly instance: ec2.Instance;
+  public readonly endpoint: string;
+
+  constructor(scope: Construct, id: string, props: PgBouncerProps) {
+    super(scope, id);
+
+    const {
+      vpc,
+      database,
+      usePublicSubnet = false,
+      instanceType = ec2.InstanceType.of(
+        ec2.InstanceClass.T3,
+        ec2.InstanceSize.MICRO,
+      ),
+      pgBouncerConfig = {
+        poolMode: "transaction",
+        maxClientConn: 200,
+        defaultPoolSize: 5,
+        minPoolSize: 0,
+        reservePoolSize: 5,
+        reservePoolTimeout: 5,
+        maxDbConnections: 40,
+        maxUserConnections: 40,
+      },
+    } = props;
+
+    // Create role for PgBouncer instance to enable writing to CloudWatch
+    const role = new iam.Role(this, "InstanceRole", {
```

Comment:
Maybe we could add a role_name/description to help group resources?

Reply:
This is the name that the role gets right now: `MAAP-STAC-test-pgSTAC-pgbouncerInstanceRoleBD5FA6DE-jCb7luVJ5MNH`

Do you want something more descriptive than that? I assume we wouldn't want to create a role that gets used by both `test` and `dev` since that would create a dependency between the two deployments.

Reply:
Ah, it inherited buildStack name. That's probably fine. Defining roleName would cut it down a little bit but it's really just a minor detail